File: python-client/src/crypto/key_migration.py
import os
import json
import base64
import socket
import logging
import time
import shutil
from pathlib import Path
from datetime import datetime

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding, rsa
from cryptography.exceptions import InvalidSignature
import hashlib
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import padding, utils

from crypto.keys import generate_keypair, save_private_key, save_public_key, load_private_key, load_public_key

logger = logging.getLogger(__name__)


class KeyMigration:
    """Handles the process of migrating to a new cryptographic key."""

    def __init__(self, peer_id, contact_manager, authentication):
        """Initialize with peer ID and managers for contacts and authentication."""
        self.peer_id = peer_id
        self.contact_manager = contact_manager
        self.authentication = authentication
        self.storage_path = Path.home() / '.p2p-share' / 'keys'

    def initiate_migration(self):
        """
        Start the key migration process.

        Returns:
            dict: Information about the migration status
        """
        try:
            # Create backup directory
            backup_dir = self.storage_path / 'backup' / \
                datetime.now().strftime('%Y%m%d_%H%M%S')
            backup_dir.mkdir(parents=True, exist_ok=True)

            # Backup current keys
            private_key_path = self.storage_path / 'private.pem'
            public_key_path = self.storage_path / 'public.pem'

            if private_key_path.exists() and public_key_path.exists():
                # Backup the keys
                shutil.copy2(private_key_path, backup_dir / 'private.pem')
                shutil.copy2(public_key_path, backup_dir / 'public.pem')

                # Store old keys
                self.old_private_key = load_private_key(private_key_path)
                self.old_public_key = load_public_key(public_key_path)

                # Generate new key pair
                self.new_private_key, self.new_public_key = generate_keypair()

                # Store the new public key in PEM format for sharing
                self.new_public_key_pem = self.new_public_key.public_bytes(
                    encoding=serialization.Encoding.PEM,
                    format=serialization.PublicFormat.SubjectPublicKeyInfo
                ).decode('utf-8')

                return {
                    "status": "ready",
                    "message": "Key migration initiated. Ready to notify contacts."
                }
            else:
                return {
                    "status": "error",
                    "message": "Current keys not found. Cannot initiate migration."
                }

        except Exception as e:
            logger.error(f"Error initiating key migration: {e}")
            return {
                "status": "error",
                "message": f"Error initiating key migration: {e}"
            }

    def notify_contacts(self):
        """
        Notify all trusted contacts about the key migration.

        Returns:
            dict: Information about the notification status
        """
        try:
            # Get all trusted contacts
            contacts = self.contact_manager.get_all_trusted_contacts()

            if not contacts:
                return {
                    "status": "warning",
                    "message": "No trusted contacts to notify."
                }

            # Create migration message
            migration_time = time.time()
            message = f"KEY_MIGRATION:{self.peer_id}:{migration_time}"

            # Sign the message with the old private key
            signature = self._sign_message(message)

            # Create notification object
            notification = {
                "peer_id": self.peer_id,
                "old_public_key": self.authentication.get_public_key_pem(),
                "new_public_key": self.new_public_key_pem,
                "signature": signature,
                "migration_time": migration_time
            }

            # Convert to JSON
            notification_json = json.dumps(notification)

            # Send notification to all contacts
            success_count = 0
            failure_count = 0

            print(f"Notifying {len(contacts)} contacts about key migration...")

            for peer_id, contact in contacts.items():
                try:
                    # Extract host and port
                    addr = contact.get("address", "")
                    if not addr or ":" not in addr:
                        logger.warning(
                            f"Invalid address for contact {peer_id}")
                        failure_count += 1
                        continue

                    host, port_str = addr.split(":")
                    port = int(port_str)

                    # Connect to the peer
                    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                    sock.settimeout(5)  # 5 second timeout
                    sock.connect((host, port))

                    # Send notification
                    sock.sendall(
                        f"MIGRATE_KEY:{notification_json}".encode('utf-8'))

                    # Wait for acknowledgment (with timeout)
                    sock.settimeout(10)  # 10 second timeout for response
                    try:
                        response = sock.recv(4096).decode('utf-8')
                        if response.startswith("MIGRATE_KEY_ACK:"):
                            ack_data = json.loads(response.split(":", 1)[1])
                            status = ack_data.get("status")

                            if status == "accepted":
                                print(
                                    f"Contact {peer_id} accepted key migration")
                                success_count += 1
                            else:
                                print(
                                    f"Contact {peer_id} rejected key migration: {status}")
                                failure_count += 1
                        else:
                            print(
                                f"Unexpected response from {peer_id}: {response[:20]}")
                            failure_count += 1
                    except socket.timeout:
                        print(f"No response from {peer_id} (timeout)")
                        failure_count += 1

                    sock.close()

                except Exception as e:
                    logger.error(f"Error notifying contact {peer_id}: {e}")
                    failure_count += 1

            return {
                "status": "complete",
                "message": f"Notification complete. Success: {success_count}, Failures: {failure_count}",
                "success_count": success_count,
                "failure_count": failure_count
            }

        except Exception as e:
            logger.error(f"Error notifying contacts: {e}")
            return {
                "status": "error",
                "message": f"Error notifying contacts: {e}"
            }

    def complete_migration(self):
        """
        Complete the key migration by saving the new keys.

        Returns:
            dict: Information about the completion status
        """
        try:
            # Save the new keys
            private_key_path = self.storage_path / 'private.pem'
            public_key_path = self.storage_path / 'public.pem'

            save_private_key(self.new_private_key, private_key_path)
            save_public_key(self.new_public_key, public_key_path)

            print("✅ New keys saved successfully")

            # Reload the keys in the authentication system
            self.authentication.load_keys()

            return {
                "status": "complete",
                "message": "Key migration completed successfully."
            }

        except Exception as e:
            logger.error(f"Error completing key migration: {e}")
            return {
                "status": "error",
                "message": f"Error completing key migration: {e}"
            }

    def _sign_message(self, message):
        """Sign a message with the old private key."""
        if not hasattr(self, 'old_private_key'):
            raise ValueError("Old private key not available")

        # Convert message to bytes if it's a string
        if isinstance(message, str):
            message = message.encode('utf-8')

        # Create signature
        signature = self.old_private_key.sign(
            message,
            padding.PSS(
                mgf=padding.MGF1(hashes.SHA256()),
                salt_length=padding.PSS.MAX_LENGTH
            ),
            hashes.SHA256()
        )

        # Return base64 encoded signature
        return base64.b64encode(signature).decode('utf-8')


def handle_key_migration(message, addr, conn, contact_manager, authentication):
    """
    Handle a key migration message from a peer.

    Args:
        message: The migration message
        addr: The peer's address
        conn: The connection socket
        contact_manager: The contact manager
        authentication: The authentication system

    Returns:
        bool: True if the migration was handled successfully
    """
    try:
        # Parse the migration notification
        notification = json.loads(message)

        peer_id = notification.get("peer_id")
        old_public_key = notification.get("old_public_key")
        new_public_key = notification.get("new_public_key")
        signature = notification.get("signature")
        migration_time = notification.get("migration_time")

        # Add more detailed logging for debugging
        message_text = f"KEY_MIGRATION:{peer_id}:{migration_time}"
        print(f"DEBUG PYTHON - Message to verify: {message_text}")
        print(
            f"DEBUG PYTHON - Message bytes (hex): {message_text.encode('utf-8').hex()}")
        print(f"DEBUG PYTHON - Base64 signature: {signature}")
        print(
            f"DEBUG PYTHON - Raw signature (hex): {base64.b64decode(signature).hex()}")
        print(
            f"DEBUG PYTHON - Public key first 50 chars: {old_public_key[:50]}...")

        if not peer_id or not old_public_key or not new_public_key or not signature or not migration_time:
            logger.error("Invalid key migration notification")
            send_migration_ack(conn, authentication.peer_id, "invalid_format")
            return False

        print(f"\nReceived key migration notification from peer {peer_id}")

        # Check if this is a trusted contact
        contact = contact_manager.get_trusted_contact(peer_id)
        if not contact:
            logger.error(f"Received key migration from unknown peer {peer_id}")
            send_migration_ack(conn, authentication.peer_id, "untrusted_peer")
            return False

        # Verify the old public key matches our stored public key
        if contact.get("public_key") != old_public_key:
            logger.error(
                f"Old public key doesn't match stored key for peer {peer_id}")
            send_migration_ack(conn, authentication.peer_id, "key_mismatch")
            return False

        # Verify signature
        message_text = f"KEY_MIGRATION:{peer_id}:{migration_time}"
        if not verify_migration_signature(message_text, signature, old_public_key):
            logger.error(
                f"Invalid signature in key migration from peer {peer_id}")
            send_migration_ack(conn, authentication.peer_id,
                               "invalid_signature")
            return False

        # Notify the user and get consent
        print(f"Peer {peer_id} is migrating to a new key.")
        print("Do you want to accept this key migration? (y/n): ", end="", flush=True)

        consent = input().lower().strip()

        if consent != 'y':
            print(f"Rejected key migration from peer {peer_id}")
            send_migration_ack(conn, authentication.peer_id, "rejected")
            return False

        # Update the contact with the new public key
        contact["public_key"] = new_public_key
        contact["verified_at"] = migration_time
        contact["last_seen"] = time.time()
        contact_manager.save_contacts()

        print(f"Updated public key for peer {peer_id}")

        # Send acknowledgment
        send_migration_ack(conn, authentication.peer_id, "accepted")

        return True

    except Exception as e:
        logger.error(f"Error handling key migration: {e}")
        try:
            send_migration_ack(
                conn, authentication.peer_id if authentication else "unknown", "error")
        except:
            pass
        return False


def verify_migration_signature(message, signature_b64, public_key_pem):
    try:
        # Convert message to bytes if it's a string
        if isinstance(message, str):
            message_bytes = message.encode('utf-8')
        else:
            message_bytes = message

        print(f"DEBUG PYTHON - Raw message bytes (hex): {message_bytes.hex()}")

        # Calculate the hash like Go does
        message_hash = hashlib.sha256(message_bytes).digest()
        print(f"DEBUG PYTHON - Message hash (hex): {message_hash.hex()}")

        # Decode the signature
        signature = base64.b64decode(signature_b64)
        print(f"DEBUG PYTHON - Signature length: {len(signature)}")

        # Load the public key
        public_key = serialization.load_pem_public_key(
            public_key_pem.encode('utf-8')
        )

        # Try both methods to verify
        try:
            # Method 1: Original method (expecting the verify function to hash)
            print("DEBUG PYTHON - Trying verification method 1 (internal hashing)")
            public_key.verify(
                signature,
                message_bytes,
                padding.PSS(
                    mgf=padding.MGF1(hashes.SHA256()),
                    salt_length=padding.PSS.MAX_LENGTH
                ),
                hashes.SHA256()
            )
            print("DEBUG PYTHON - Method 1 succeeded!")
            return True
        except Exception as e1:
            print(f"DEBUG PYTHON - Method 1 failed: {e1}")

            try:
                # Method 2: Pre-hashed method (matching Go's approach)
                print("DEBUG PYTHON - Trying verification method 2 (pre-hashed)")
                public_key.verify(
                    signature,
                    message_hash,
                    padding.PSS(
                        mgf=padding.MGF1(hashes.SHA256()),
                        salt_length=padding.PSS.MAX_LENGTH
                    ),
                    utils.Prehashed(hashes.SHA256())
                )
                print("DEBUG PYTHON - Method 2 succeeded!")
                return True
            except Exception as e2:
                print(f"DEBUG PYTHON - Method 2 failed: {e2}")
                raise e2

    except InvalidSignature:
        logger.warning("Invalid signature in migration verification")
        return False
    except Exception as e:
        logger.error(f"Error verifying migration signature: {e}")
        return False


def send_migration_ack(conn, peer_id, status):
    """
    Send an acknowledgment for a key migration notification.

    Args:
        conn: The connection socket
        peer_id: Our peer ID
        status: The acknowledgment status
    """
    try:
        ack = {
            "peer_id": peer_id,
            "status": status
        }

        ack_json = json.dumps(ack)
        conn.sendall(f"MIGRATE_KEY_ACK:{ack_json}".encode('utf-8'))
    except Exception as e:
        logger.error(f"Error sending migration acknowledgment: {e}")

File: go-client/internal/crypto/key_migration.go
package crypto

import (
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"io"
	"net"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// KeyMigration handles the process of migrating to a new key
type KeyMigration struct {
	PeerID         string
	OldPrivateKey  *rsa.PrivateKey
	OldPublicKey   *rsa.PublicKey
	NewPrivateKey  *rsa.PrivateKey
	NewPublicKey   *rsa.PublicKey
	ContactManager *ContactManager
}

// MigrationNotification represents the notification message for key migration
type MigrationNotification struct {
	PeerID        string  `json:"peer_id"`
	OldPublicKey  string  `json:"old_public_key"`
	NewPublicKey  string  `json:"new_public_key"`
	Signature     string  `json:"signature"` // Signature created with old private key
	MigrationTime float64 `json:"migration_time"`
}

// InitiateMigration starts the process of migrating to a new key
func InitiateMigration(peerID string, contactManager *ContactManager) (*KeyMigration, error) {
	// Check if we have valid parameters
	if peerID == "" || contactManager == nil {
		return nil, fmt.Errorf("invalid peer ID or contact manager")
	}

	// Get paths for keys
	homeDir, err := os.UserHomeDir()
	if err != nil {
		return nil, fmt.Errorf("error getting home directory: %v", err)
	}

	keysDir := filepath.Join(homeDir, ".p2p-share", "keys")
	oldPrivateKeyPath := filepath.Join(keysDir, "private.pem")
	oldPublicKeyPath := filepath.Join(keysDir, "public.pem")

	// Create backup directory
	backupDir := filepath.Join(keysDir, "backup", time.Now().Format("20060102_150405"))
	err = os.MkdirAll(backupDir, 0755)
	if err != nil {
		return nil, fmt.Errorf("error creating backup directory: %v", err)
	}

	// Load old keys
	oldPrivateKey, err := LoadPrivateKey(oldPrivateKeyPath)
	if err != nil {
		return nil, fmt.Errorf("error loading old private key: %v", err)
	}

	oldPublicKey, err := LoadPublicKey(oldPublicKeyPath)
	if err != nil {
		return nil, fmt.Errorf("error loading old public key: %v", err)
	}

	// Backup old keys
	oldPrivateKeyBackupPath := filepath.Join(backupDir, "private.pem")
	oldPublicKeyBackupPath := filepath.Join(backupDir, "public.pem")

	err = copyFile(oldPrivateKeyPath, oldPrivateKeyBackupPath)
	if err != nil {
		return nil, fmt.Errorf("error backing up private key: %v", err)
	}

	err = copyFile(oldPublicKeyPath, oldPublicKeyBackupPath)
	if err != nil {
		return nil, fmt.Errorf("error backing up public key: %v", err)
	}

	// Generate new keys
	newPrivateKey, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		return nil, fmt.Errorf("error generating new key pair: %v", err)
	}

	// Create migration object
	migration := &KeyMigration{
		PeerID:         peerID,
		OldPrivateKey:  oldPrivateKey,
		OldPublicKey:   oldPublicKey,
		NewPrivateKey:  newPrivateKey,
		NewPublicKey:   &newPrivateKey.PublicKey,
		ContactManager: contactManager,
	}

	return migration, nil
}

// CompleteMigration finalizes the key migration process
func (km *KeyMigration) CompleteMigration() error {
	// Get paths for keys
	homeDir, err := os.UserHomeDir()
	if err != nil {
		return fmt.Errorf("error getting home directory: %v", err)
	}

	keysDir := filepath.Join(homeDir, ".p2p-share", "keys")
	privateKeyPath := filepath.Join(keysDir, "private.pem")
	publicKeyPath := filepath.Join(keysDir, "public.pem")

	// Save new keys
	err = savePrivateKey(km.NewPrivateKey, privateKeyPath)
	if err != nil {
		return fmt.Errorf("error saving new private key: %v", err)
	}

	err = savePublicKey(km.NewPublicKey, publicKeyPath)
	if err != nil {
		return fmt.Errorf("error saving new public key: %v", err)
	}

	fmt.Println("✅ New keys generated and saved successfully")
	return nil
}

// savePrivateKey saves a private key to a file
func savePrivateKey(key *rsa.PrivateKey, path string) error {
	file, err := os.Create(path)
	if err != nil {
		return err
	}
	defer file.Close()

	return pem.Encode(file, &pem.Block{
		Type:  "RSA PRIVATE KEY",
		Bytes: x509.MarshalPKCS1PrivateKey(key),
	})
}

// savePublicKey saves a public key to a file
func savePublicKey(key *rsa.PublicKey, path string) error {
	file, err := os.Create(path)
	if err != nil {
		return err
	}
	defer file.Close()

	pubBytes, err := x509.MarshalPKIXPublicKey(key)
	if err != nil {
		return err
	}

	return pem.Encode(file, &pem.Block{
		Type:  "PUBLIC KEY",
		Bytes: pubBytes,
	})
}

// GetPublicKeyPEM returns the PEM encoded public key
func GetPublicKeyPEM(key *rsa.PublicKey) string {
	pubKeyBytes, err := x509.MarshalPKIXPublicKey(key)
	if err != nil {
		return ""
	}

	pubKeyPEM := pem.EncodeToMemory(&pem.Block{
		Type:  "PUBLIC KEY",
		Bytes: pubKeyBytes,
	})

	return string(pubKeyPEM)
}

// SignMessage signs a message with the private key
func SignMessage(key *rsa.PrivateKey, message string) (string, error) {
	fmt.Printf("DEBUG GO - Raw message bytes (hex): %x\n", []byte(message))
	msgHash := sha256.Sum256([]byte(message))
	fmt.Printf("DEBUG GO - Message hash (hex): %x\n", msgHash[:])

	signature, err := rsa.SignPSS(rand.Reader, key, crypto.SHA256, msgHash[:], nil)
	if err != nil {
		return "", err
	}

	return base64.StdEncoding.EncodeToString(signature), nil
}

// NotifyContacts notifies all trusted contacts about the key migration
func (km *KeyMigration) NotifyContacts() error {
	// Get the public keys in PEM format
	oldPubKeyPEM := GetPublicKeyPEM(km.OldPublicKey)
	newPubKeyPEM := GetPublicKeyPEM(km.NewPublicKey)

	// Get the list of trusted contacts
	contacts := km.ContactManager.GetAllTrustedContacts()

	// Create notification message
	migrationTime := float64(time.Now().Unix())
	// Change this line in NotifyContacts or wherever you create the message
	message := fmt.Sprintf("KEY_MIGRATION:%s:%d", km.PeerID, int64(migrationTime))

	// Add more detailed logging for debugging

	fmt.Printf("DEBUG GO - Message to sign: %s\n", message)

	// Sign the message with the old private key
	signature, err := SignMessage(km.OldPrivateKey, message)
	if err != nil {
		return fmt.Errorf("error signing migration message: %v", err)
	}

	fmt.Printf("DEBUG GO - Base64 signature: %s\n", signature)
	fmt.Printf("DEBUG GO - Old public key: %s\n", GetPublicKeyPEM(km.OldPublicKey))

	// Create notification payload
	notification := MigrationNotification{
		PeerID:        km.PeerID,
		OldPublicKey:  oldPubKeyPEM,
		NewPublicKey:  newPubKeyPEM,
		Signature:     signature,
		MigrationTime: migrationTime,
	}

	notificationJSON, err := json.Marshal(notification)
	if err != nil {
		return fmt.Errorf("error encoding notification: %v", err)
	}

	// Send notification to all contacts
	fmt.Printf("Notifying %d contacts about key migration\n", len(contacts))

	for peerID, contact := range contacts {
		fmt.Printf("Notifying contact %s at %s\n", peerID, contact.Address)

		// Parse address
		parts := strings.Split(contact.Address, ":")
		if len(parts) != 2 {
			fmt.Printf("Invalid address format for contact %s: %s\n", peerID, contact.Address)
			continue
		}

		host := parts[0]
		port := parts[1]

		// Connect to the peer
		conn, err := net.Dial("tcp", net.JoinHostPort(host, port))
		if err != nil {
			fmt.Printf("Error connecting to contact %s: %v\n", peerID, err)
			continue
		}

		// Send notification
		_, err = conn.Write([]byte(fmt.Sprintf("MIGRATE_KEY:%s", notificationJSON)))
		if err != nil {
			fmt.Printf("Error sending notification to contact %s: %v\n", peerID, err)
		} else {
			fmt.Printf("Successfully notified contact %s\n", peerID)
		}

		conn.Close()
	}

	return nil
}

// HandleMigrateKeyMessage processes a key migration notification from a peer
func HandleMigrateKeyMessage(conn net.Conn, addr string, payload string) error {
	// Parse the notification
	var notification MigrationNotification
	err := json.Unmarshal([]byte(payload), &notification)
	if err != nil {
		return fmt.Errorf("error parsing migration notification: %v", err)
	}

	fmt.Printf("\nReceived key migration notification from peer %s\n", notification.PeerID)

	// Verify notification signature
	message := fmt.Sprintf("KEY_MIGRATION:%s:%f", notification.PeerID, notification.MigrationTime)

	// Get the contact
	contact, exists := contactManager.GetTrustedContact(notification.PeerID)
	if !exists {
		return fmt.Errorf("peer %s is not a trusted contact", notification.PeerID)
	}

	// Verify the old public key matches our stored public key
	if contact.PublicKey != notification.OldPublicKey {
		return fmt.Errorf("old public key in notification does not match our stored key for peer %s", notification.PeerID)
	}

	// Verify signature using old public key
	oldPubKey, err := parsePublicKeyPEM(notification.OldPublicKey)
	if err != nil {
		return fmt.Errorf("error parsing old public key: %v", err)
	}

	signatureBytes, err := base64.StdEncoding.DecodeString(notification.Signature)
	if err != nil {
		return fmt.Errorf("error decoding signature: %v", err)
	}

	msgHash := sha256.Sum256([]byte(message))
	err = rsa.VerifyPSS(oldPubKey, crypto.SHA256, msgHash[:], signatureBytes, nil)
	if err != nil {
		return fmt.Errorf("invalid signature in migration notification: %v", err)
	}

	// Valid notification - update the contact
	fmt.Printf("Verified key migration from peer %s\n", notification.PeerID)
	fmt.Printf("Do you want to accept this key migration? (y/n): ")
	var consent string
	fmt.Scanln(&consent)

	if strings.ToLower(consent) != "y" {
		fmt.Printf("Key migration from peer %s rejected\n", notification.PeerID)
		return nil
	}

	// Update the contact with the new public key
	contact.PublicKey = notification.NewPublicKey
	contact.VerifiedAt = notification.MigrationTime
	contact.LastSeen = float64(time.Now().Unix())

	contactManager.SaveContacts()

	fmt.Printf("Updated public key for peer %s\n", notification.PeerID)

	// Send acknowledgment
	ack := map[string]interface{}{
		"peer_id": authentication.PeerID,
		"status":  "accepted",
	}

	ackJSON, err := json.Marshal(ack)
	if err == nil {
		conn.Write([]byte(fmt.Sprintf("MIGRATE_KEY_ACK:%s", ackJSON)))
	}

	return nil
}

// parsePublicKeyPEM parses a PEM encoded public key
func parsePublicKeyPEM(pemStr string) (*rsa.PublicKey, error) {
	block, _ := pem.Decode([]byte(pemStr))
	if block == nil || block.Type != "PUBLIC KEY" {
		return nil, fmt.Errorf("failed to decode PEM block containing public key")
	}

	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return nil, err
	}

	pubKey, ok := pub.(*rsa.PublicKey)
	if !ok {
		return nil, fmt.Errorf("not an RSA public key")
	}

	return pubKey, nil
}

// copyFile copies a file from src to dst
func copyFile(src, dst string) error {
	sourceFile, err := os.Open(src)
	if err != nil {
		return err
	}
	defer sourceFile.Close()

	destFile, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer destFile.Close()

	_, err = io.Copy(destFile, sourceFile)
	if err != nil {
		return err
	}

	return nil
}
